### ISSUE TITLE
rustc 1.82.0 doesn't add LIBPATH when invoking link.exe

### ISSUE THREAD

**Issue by nandin-borjigin** [rust-lang/rust#132016]
Rust-to-rust dynamic linking (through import library) on Windows fails with toolchain version 1.82.0. 

The build log shows that 1.82.0 is not adding any /LIBPATH arguments when invoking link.exe.

I tried this code: [nandin-borjigin/rustc-1.82.0-repro](https://github.com/nandin-borjigin/rustc-1.82.0-repro)

![Image](https://github.com/user-attachments/assets/cafb5b8c-4229-4a74-8ff8-8821d320fae4)

I expected to see this happen: *Successful build*

Instead, this happened: *LINK : fatal error LNK1181: cannot open input file 'mylib.dll.lib'*

![Image](https://github.com/user-attachments/assets/a9f0bdf1-f49d-4667-bbb1-5325474fbf9a)

### Meta

`rustc --version --verbose`:
```
rustc 1.82.0 (f6e511eec 2024-10-15)
binary: rustc
commit-hash: f6e511eec7342f59a25f7c0534f1dbea00d01b14
commit-date: 2024-10-15
host: x86_64-pc-windows-msvc
release: 1.82.0
LLVM version: 19.1.1
```

<details><summary>Backtrace</summary>
<p>

</p>
</details>

### Possibly related

- #131995

**Comment by ChrisDenton** [comment]
cc @petrochenkov, I believe there were some changes to linking with msvc?

**Comment by bjorn3** [comment]
I don't think this was supposed to work out of the box. Either you should make my_lib a cdylib (requiring you to build the cdylib from within the build script rather than specifying it as dependency, at least until artifact dependencies are stable) and link it like a C library (which includes adding it to the linker search path from a build script) or you should link it like a regular rust library and avoid `#[link(name = "my_lib.dll")] extern "C" {}`.

**Comment by petrochenkov** [comment]
This is likely from https://github.com/rust-lang/rust/pull/128370.

**Comment by nandin-borjigin** [comment]
Hi @bjorn3 , thanks for the input. I have a few follow up questions:
1. As per [The Rust Reference/Linkage](https://doc.rust-lang.org/reference/linkage.html), the difference between `dylib` and `cdylib` seems to be whether the produced library is consumable in Rust or other languages. If `mylib` is only going to be used in Rust, do we need to use `cdylib`?
2. If `mylib` is changed to `cdylib`, the compiler would warn saying that `mylib` provides no linkable target. This should be the reason why you mentioned `mylib` should be built form the build script in this case. I wonder how I should properly do that in build script? Will triggering `cargo build ../mylib` be sufficient? or do I have to take care of (debug|release) profiles and crate features as well?

**Comment by bjorn3** [comment]
>  mylib is only going to be used in Rust, do we need to use cdylib?

If you want to use it as regular dependency you should make it a dylib.

> Will triggering cargo build ../mylib be sufficient? or do I have to take care of (debug|release) profiles and crate features as well?

You probably want to forward the build profile. Also make sure to set `--target-dir` to some subdirectory of `OUT_DIR`. If you have any relevant crate features you will also need to forward them.

By the way what do you want to achieve in the end for which you see dylibs as solution?

**Comment by nandin-borjigin** [comment]
> By the way what do you want to achieve in the end for which you see dylibs as solution?

`mylib` implements an `Allocator`, which is configured as the `GlobalAllocator` in `app`. But we don't want to get into a recursive dead lock where the allocations in `mylib::Allocator` is also driven by `mylib::Allocator` itself.

**Comment by bjorn3** [comment]
You can call into the system allocator from mylib using `std::alloc::System` or you can directly use mmap.

**Comment by jieyouxu** [comment]
Triage: I believe the change is intentional.
